### PR TITLE
Balder/sameelement in streaming

### PR DESCRIPTION
**Pull request by baldersheim**
@geirst PR
@vekterli You might be interested in the first 2 commits
It can be review in sequence or all together.

### CODE REVIEW

File: searchlib/src/vespa/searchlib/query/query.cpp [hunk 1]
```diff
@@ -28,85 +28,77 @@ const HitList & QueryConnector::evaluateHits(HitList & hl) const
...
-    const QueryNode & qn = **it;
-    size_t t = qn.depth();
-    if (t > d)
-      d = t;
-  }
-  return d+1;
+    size_t d(0);
+    for(const auto & node : *this) {
+        size_t t = node->depth();
+        if (t > d)
```
Comment (geirst) [inline searchlib/src/vespa/searchlib/query/query.cpp]: Please add "{" and "}"

File: searchlib/src/vespa/searchlib/query/query.cpp [hunk 2]
```diff
@@ -153,13 +145,63 @@ bool EquivQueryNode::evaluate() const
...
+    auto curr = static_cast<const QueryTerm *> (&(*(*this)[currMatchCount]));
+    bool exhausted( curr->evaluateHits(tmpHL).empty());
+    for (; !exhausted; ) {
+        auto next = static_cast<const QueryTerm &>(*(*this)[currMatchCount+1]);
+        unsigned int & currIndex = indexVector[currMatchCount];
+        unsigned int & nextIndex = indexVector[currMatchCount+1];
+
+        const auto & currHit = curr->evaluateHits(tmpHL)[currIndex];
+        uint32_t currElemId = currHit.elemId();
+        uint32_t curContext = currHit.context();
```
Comment (geirst) [inline searchlib/src/vespa/searchlib/query/query.cpp]: Please rename curContext -> currContext

File: searchlib/src/tests/query/query-old.cpp [hunk 3]
```diff
@@ -651,4 +651,80 @@ TEST("require that we do not break the stack on bad query") {
     EXPECT_FALSE(term.isValid());
 }
 
+TEST("testPhraseEvaluate") {
```
Comment (geirst) [inline searchlib/src/tests/query/query-old.cpp]: Please rename test name

File: searchlib/src/tests/query/query-old.cpp [hunk 4]
```diff
@@ -651,4 +651,80 @@ TEST("require that we do not break the stack on bad query") {
...
+    {
+        builder.addStringTerm("a", "f1", 0, Weight(0));
+        builder.addStringTerm("b", "f2", 1, Weight(0));
+        builder.addStringTerm("c", "f3", 2, Weight(0));
+    }
+    Node::UP node = builder.build();
+    vespalib::string stackDump = StackDumpCreator::create(*node);
+    QueryNodeResultFactory empty;
+    Query q(empty, stackDump);
+    SameElementQueryNode * p = dynamic_cast<SameElementQueryNode *>(&q.getRoot());
```
Comment (geirst) [inline searchlib/src/tests/query/query-old.cpp]: 'p' seems like a strange variable name here. Consider 'node' instead

File: searchlib/src/tests/query/query-old.cpp [hunk 5]
```diff
@@ -651,4 +651,80 @@ TEST("require that we do not break the stack on bad query") {
...
+    EXPECT_TRUE(p != nullptr);
+    EXPECT_EQUAL("field", p->getIndex());
+    EXPECT_EQUAL(3u, p->size());
+    EXPECT_TRUE(dynamic_cast<const QueryTerm *>((*p)[0].get()) != nullptr);
```
Comment (geirst) [inline searchlib/src/tests/query/query-old.cpp]: Consider helper function for the dynamic cast and name check

File: searchlib/src/tests/query/query-old.cpp [hunk 6]
```diff
@@ -651,4 +651,80 @@ TEST("require that we do not break the stack on bad query") {
...
+    terms[1]->add(0, 1, 2, 1);
+    terms[1]->add(0, 1, 4, 1);
+    terms[1]->add(0, 1, 5, 1);
+    terms[1]->add(0, 1, 6, 1);
+
+    terms[2]->add(0, 2, 0, 1);
+    terms[2]->add(0, 2, 2, 1);
+    terms[2]->add(0, 2, 4, 1);
+    terms[2]->add(0, 2, 5, 1);
+    terms[2]->add(0, 2, 6, 1);
```
Comment (geirst) [inline searchlib/src/tests/query/query-old.cpp]: Perhaps we should test with different wordpos and weight to check how this end up in the Hit.

**Review by geirst — COMMENTED** [review]
Overall looks good, some comments to look at

**Comment by geirst** [comment]
@baldersheim also note that build has failed

File: document/src/vespa/document/fieldvalue/iteratorhandler.cpp [hunk 7]
```diff
@@ -18,11 +25,13 @@ IteratorHandler::handleComplex(const FieldValue & fv) {
 }
 void
 IteratorHandler::handleCollectionStart(const FieldValue & fv) {
+    _arrayIndexStack.push_back(0);
     onCollectionStart(Content(fv, getWeight()));
 }
 void
 IteratorHandler::handleCollectionEnd(const FieldValue & fv) {
     onCollectionEnd(Content(fv, getWeight()));
+    _arrayIndexStack.resize(_arrayIndexStack.size() - 1);
```
[the remark this answers is not in the document]
Comment (baldersheim) [reply document/src/vespa/document/fieldvalue/iteratorhandler.cpp]: Done

File: searchlib/src/vespa/searchlib/query/query.cpp [hunk 8]
```diff
@@ -181,14 +182,23 @@ const HitList & PhraseQueryNode::evaluateHits(HitList & hl) const
       const QueryTerm & next = static_cast<const QueryTerm &>(*(*this)[currPhraseLen+1]);
       unsigned int & currIndex = indexVector[currPhraseLen];
       unsigned int & nextIndex = indexVector[currPhraseLen+1];
-      const HitList & nextHL = next.evaluateHits(tmpHL);
 
       size_t firstPosition = curr->evaluateHits(tmpHL)[currIndex].pos();
+      uint32_t currElemId = curr->evaluateHits(tmpHL)[currIndex].elemId();
```
[the remark this answers is not in the document]
Comment (baldersheim) [reply searchlib/src/vespa/searchlib/query/query.cpp]: Done in later commit.

File: searchlib/src/vespa/searchlib/query/query.cpp [hunk 9]
```diff
@@ -181,14 +182,23 @@ const HitList & PhraseQueryNode::evaluateHits(HitList & hl) const
...
+      uint32_t curContext = curr->evaluateHits(tmpHL)[currIndex].context();
```
[the remark this answers is not in the document]
Comment (baldersheim) [reply searchlib/src/vespa/searchlib/query/query.cpp]: Done

File: searchlib/src/vespa/searchlib/query/query.cpp (hunk 1 above)
Comment (baldersheim) [reply searchlib/src/vespa/searchlib/query/query.cpp]: Done

File: searchlib/src/tests/query/query-old.cpp (hunk 3 above)
Comment (baldersheim) [reply searchlib/src/tests/query/query-old.cpp]: Done

File: searchlib/src/tests/query/query-old.cpp (hunk 4 above)
Comment (baldersheim) [reply searchlib/src/tests/query/query-old.cpp]: sameElem

File: searchlib/src/tests/query/query-old.cpp (hunk 5 above)
Comment (baldersheim) [reply searchlib/src/tests/query/query-old.cpp]: Done

File: searchlib/src/tests/query/query-old.cpp (hunk 6 above)
Comment (baldersheim) [reply searchlib/src/tests/query/query-old.cpp]: Done

File: searchlib/src/tests/query/query-old.cpp [hunk 10]
```diff
@@ -681,48 +685,48 @@ TEST("testPhraseEvaluate") {
...
     EXPECT_EQUAL(2u, hits[2].context());
     EXPECT_EQUAL(4u, hits[2].elemId());
-    EXPECT_EQUAL(1,  hits[2].weight());
+    EXPECT_EQUAL(150,  hits[2].weight());
 
     EXPECT_EQUAL(0u, hits[3].wordpos());
     EXPECT_EQUAL(2u, hits[3].context());
     EXPECT_EQUAL(5u, hits[3].elemId());
-    EXPECT_EQUAL(1,  hits[3].weight());
+    EXPECT_EQUAL(160,  hits[3].weight());
```
Comment (geirst) [inline searchlib/src/tests/query/query-old.cpp]: Nitpick: Based on the current test data it is hard to tell whether resulting weight is max or the last found. Would be nice to change test data to differentiate.

**Review by geirst — COMMENTED** [review]
👍

**Pull request merged**